dex/src/state.rs

```diff
@@ -2686,8 +2680,168 @@ impl State {
     }
 
     #[cfg(feature = "program")]
-    fn process_send_take(_args: account_parser::SendTakeArgs) -> DexResult {
-        unimplemented!()
+    fn process_send_take(args: account_parser::SendTakeArgs) -> DexResult {
+        let account_parser::SendTakeArgs {
+            instruction,
+            signer,
+            mut event_q,
+            mut req_q,
+            mut order_book_state,
+            coin_wallet,
+            pc_wallet,
+            coin_vault,
+            pc_vault,
+            spl_token_program,
+            fee_tier,
+            vault_signer,
+        } = args;
+
+        let order_id = req_q.gen_order_id(instruction.limit_price.get(), instruction.side);
+        let native_pc_qty_locked = match instruction.side {
+            Side::Bid => {
+                let lock_qty_native = instruction.max_native_pc_qty_including_fees;
+                Some(lock_qty_native)
+            }
+            Side::Ask => None,
+        };
+
+        let request = RequestView::NewOrder {
+            side: instruction.side,
+            order_type: OrderType::ImmediateOrCancel,
+            order_id,
+            fee_tier,
+            self_trade_behavior: SelfTradeBehavior::AbortTransaction,
+            // Pass dummy account - system program pubkey
+            owner: system_program::ID.to_aligned_bytes(),
+            // Pass 0
+            owner_slot: 0,
+            max_coin_qty: instruction.max_coin_qty,
+            native_pc_qty_locked,
+            client_order_id: None,
+        };
+
+        let mut limit = instruction.limit;
+        let mut proceeds = RequestProceeds::zero();
+        let _unfilled_portion = order_book_state.process_orderbook_request(
+            &request,
+            &mut event_q,
+            &mut proceeds,
+            &mut limit,
+        )?;
+
+        let market_state = order_book_state.market_state;
+
+        let coin_lot_size = market_state.coin_lot_size;
+
+        let RequestProceeds {
+            coin_unlocked: _,
+            coin_credit,
+
+            native_pc_unlocked: _,
+            native_pc_credit,
+
+            coin_debit,
+            native_pc_debit,
+        } = proceeds;
```

Comment:
maybe it's possible to sanity check that coin_unlocked and native_pc_unlocked are zero?